### PR TITLE
Whisper: Add functionality to decode without conditioning on previous text.

### PR DESCRIPTION
**Pull request by m-bain**
# What does this PR do?
Add whisper functionality to decode *without* conditioning on previous text. https://github.com/openai/whisper/blob/7858aa9c08d98f75575035ecd6481f462d66ca27/whisper/transcribe.py#L278

When condition_on_prev_text=False (in the config), the following happens:
Decoder attention is masked from the consecutive timestamps, such that each whisper segment within the chunk input does not attend to text in previous segments.

This has been shown to drastically reduce hallucination. Works for any batch_size, providing rapid decoding strategy.

Fixes # (issue)
https://github.com/huggingface/transformers/issues/21467

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@ArthurZucker

### CODE REVIEW

**Comment by ArthurZucker** [comment]
Can you run `make style` and make sure you rebase on main first? We recently merged a huge PR related to the code styling! See #21480

File: src/transformers/models/whisper/configuration_whisper.py [hunk 1]
```diff
@@ -129,14 +129,17 @@ class WhisperConfig(PretrainedConfig):
             Begin of stream token id.
         eos_token_id (`int`, *optional*, defaults to 50257):
             End of stream token id.
+        ts_start_token_id (`int`, *optional*, defaults to 50363)
```
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/configuration_whisper.py:132]: Would suggest making this more verbose in the style of `transformers` 🤗
```suggestion
        timestamp_token_id (`int`, *optional*, defaults to 50363)
```

File: src/transformers/models/whisper/modeling_whisper.py [hunk 2]
```diff
@@ -81,6 +81,33 @@ def _make_causal_mask(input_ids_shape: torch.Size, dtype: torch.dtype, past_key_
         mask = torch.cat([torch.zeros(tgt_len, past_key_values_length, dtype=dtype), mask], dim=-1)
     return mask[None, None, :, :].expand(bsz, 1, tgt_len, tgt_len + past_key_values_length)
 
+# To only condition from certain indexes
```
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/modeling_whisper.py:84]: Can remove this comment here and place it below in the docstring for the function!

File: src/transformers/models/whisper/modeling_whisper.py [hunk 3]
```diff
@@ -81,6 +81,33 @@ def _make_causal_mask(input_ids_shape: torch.Size, dtype: torch.dtype, past_key_
...
+def _make_variable_start_mask(input_ids_shape: torch.Size, dtype: torch.dtype, per_sample_start: torch.tensor,
```
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/modeling_whisper.py:85]: Do you think we could just define one function `_make_causal_mask` which takes `per_sample_start` as an optional argument? 

If this arg defaults to `None`, then we due our usual causal mask function. 

If it's not `None` and `per_sample_start.sum() > 1`, then we only condition on certain indices using this modified algorithm?

I think this way we repeat less of the code and make the control flow a bit easier to understand. WDYT?

File: src/transformers/models/whisper/modeling_whisper.py [hunk 4]
```diff
@@ -81,6 +81,33 @@ def _make_causal_mask(input_ids_shape: torch.Size, dtype: torch.dtype, past_key_
...
+                            past_key_values_length: int = 0):
+    """
+    Make causal mask used for bi-directional self-attention.
+    """
+    bsz, tgt_len = input_ids_shape
+
+    if input_ids_shape[0] != per_sample_start.shape[0]:
+        raise ValueError(f"Per sample start should be (b,n) where b is batch size {input_ids_shape[0]}, not {per_sample_start.shape[0]}")
+    # print(past_key_values_length)
```
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/modeling_whisper.py:94]: Some debugging statements here! 👀

File: src/transformers/models/whisper/modeling_whisper.py [hunk 5]
```diff
@@ -1185,6 +1225,26 @@ def forward(
...
+
+        # create causal mask
+        # [bsz, seq_len] -> [bsz, 1, tgt_seq_len, src_seq_len]
+
+        if self.config.condition_on_previous_text or torch.all(decoder_input_ids == self.config.decoder_start_token_id):
+            self.condition_start_idx = torch.zeros_like(decoder_input_ids).int()
+            self.last_is_timestamp = None
+
+        if not self.config.condition_on_previous_text:
+            timestamp_tokens: torch.Tensor = decoder_input_ids.ge(self.config.ts_start_token_id)
```
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/modeling_whisper.py:1237]: Very nicely done! Only a small nit here (we don't include typehints in-line)

```suggestion
            timestamp_tokens = decoder_input_ids.ge(self.config.ts_start_token_id)
```

**Review by sanchit-gandhi — COMMENTED** [review]
Very cool PR @m-bain! Thanks for the contribution! The modelling code looks in very good shape already. Left a few comments in the review 🤗

Could we also add a test if possible?

**Comment by ArthurZucker** [comment]
Feel free to ping me once you are done adressing all the previous comments and want another review :wink

File: src/transformers/models/whisper/modeling_whisper.py (hunk 3 above)
Comment (m-bain) [reply src/transformers/models/whisper/modeling_whisper.py:85]: makes sense!

**Comment by m-bain** [comment]
@ArthurZucker @sanchit-gandhi Thanks for the review and comments sanchit!
Unfortuntately I am pushing for ICCV/interspeech deadline for the next couple weeks so I dont have time at the moment to polish this for merge. Feel free to edit directly / use as inspo.
Btw it seems from my results that it can still occasionally repeat words (I think this is due to greedy decoding), but overall this is worth the speedup compared to beam search of 5
